### ISSUE TITLE
union copy files in specific upstream when reading

### ISSUE THREAD

**Issue by cberni** [rclone/rclone#6934]
```
[union]
type = union
action_policy = all
create_policy = all
search_policy = ff

[remote]
type = google cloud storage
service_account_file = 
```

```
RCLONE_VFS_CACHE_MODE=writes
/usr/bin/rclone mount union: /union --union-upstreams /local remote:bucket --use-mmap --disable-http2 --rc
```

we have a "union mount" of "local disk" and "remote bakend".

first found = /local 
second found = remote:bucket

- when making a "search_policy" of a file that is in /local it does not need to create cache because is local.
- when making a "search_policy" of a file that is not in /local (only in remote backend) it will be downloaded and not create mount cache because we are using --vfs-cache-mode=writes.
we know that creating --vfs-cache-mode=full will create a cache on rclone mount but that's not what we need.

the issue is: when make a "search_policy" of file that is only in "remote bakend" copy (recreate) it to "/local" upstream. This will work as a "new cache" as long as this will be kept in /local.

perhaps a way to define a specific upstream to recreate files at read time if it doesn't exist in it.
this copy (to /local again) must be done only when the file is full downloaded. When just listing files nothing to do.

**Comment by ncw** [comment]
If the vfs cache could somehow look through the union into the local backend and **not** make a copy of the file that is on the local disk, then using `--vfs-cache-mode full` would fix your problem wouldn't it?

The VFS cache would then read the local files directly from the local disk, but cache the remote files when it needed to.

Would that work for you? That is a feature many users have asked for before.

**Comment by cberni** [comment]
no. this will not work because because `--vfs-cache-mode full` will only cache into local for a few hours. And we need the file to stay into local until we say we don't need it anymore (around 30 days).

**Comment by ncw** [comment]
> no. this will not work because because `--vfs-cache-mode full` will only cache into local for a few hours. And we need the files to stay into local until we say we don't need it anymore (around 30 days).

You can control this yourself - setting `--vfs-cache-max-age 30d` for example.

    --vfs-cache-max-age duration         Max time since last access of objects in the cache (default 1h0m0s)

Or even

      --vfs-cache-max-size SizeSuffix          Max total size of objects in the cache (default off)

**Comment by cberni** [comment]
with improved VFS cache (read the local files directly from the local disk) maybe yes. But we have that memory usage problem. Every day we run `rclone rc vfs/forget` and restart the `rclone mount` service at least once a week.

we're talking about thousands of small files.
maybe a tag (ex: --search-policy-copy-local) that enabled this copy to /local it would be interesting.

**Comment by ncw** [comment]
Getting the union to write files back to the /local is sounding very much like a cache. We have a cache already though the VFS cache and I'd like to use that if possible.

If the VFS cache was working properly would you need the union at all?

I understand your workflow is writing files to the local disk which then get moved to the cloud. The union provides a unified view of all the files.

Potentially you could be just using the VFS cache `full` over the `remote:bucket` to do this? The VFS cache has parameters for removing cache files when they are too old/big and it can control the upload of modified files after a certain interval.

**Comment by cberni** [comment]
Yes. we use union (with /local) to be a large cache controlled by us (we like this option). behind we use rclone copy and move to check copies and remove old ones respectively.

if VFS cache didn't have checking (hours checking files) and memory problems. maybe we could use it. But then we would have to base it on the total size of the disk. For example an 8tb disk will use 85% of it. Could have years of cached files in some cases. So you're almost 8tb of billions of files (average of 250kb each). if a file is read refresh your cache or download it into cache again to be deleted last.

when a series of files are requested that are no longer into /local they need to be delivered for comparison and opened at least once more within the same week. This is generating a higher cost having to download directly from the remote backend twice. These files could be from 1 or 2 years ago for example.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.64
